extensions/variants_test.go

```diff
@@ -470,3 +470,330 @@ func TestResolveTypeErrorHandling(t *testing.T) {
 	require.Error(t, err)
 	assert.Contains(t, err.Error(), "mismatch in number of arguments")
 }
+
+func TestValidateConstrainedAnyTypeConsistency(t *testing.T) {
+	t.Run("non-variadic with matching types succeeds", func(t *testing.T) {
+		// func(any1, any1) with (i32, i32)
+		params := []types.FuncDefArgType{
+			&types.AnyType{Name: "any1"},
+			&types.AnyType{Name: "any1"},
+		}
+		args := []types.Type{
+			&types.Int32Type{},
+			&types.Int32Type{},
+		}
+
+		err := extensions.ValidateConstrainedAnyTypeConsistency(params, args, nil)
+		require.NoError(t, err)
+	})
+
+	t.Run("non-variadic with mismatched types fails", func(t *testing.T) {
+		// func(any1, any1) with (i32, i64)
+		params := []types.FuncDefArgType{
+			&types.AnyType{Name: "any1"},
+			&types.AnyType{Name: "any1"},
+		}
+		args := []types.Type{
+			&types.Int32Type{},
+			&types.Int64Type{},
+		}
+
+		err := extensions.ValidateConstrainedAnyTypeConsistency(params, args, nil)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "type parameter any1 cannot be both i32 and i64")
+	})
+
+	t.Run("variadic with matching types succeeds", func(t *testing.T) {
+		// func(any1, ...) with (i32, i32, i32)
+		params := []types.FuncDefArgType{
+			&types.AnyType{Name: "any1"},
+		}
+		args := []types.Type{
+			&types.Int32Type{},
+			&types.Int32Type{},
+			&types.Int32Type{},
+		}
+		variadic := &extensions.VariadicBehavior{Min: 2}
+
+		err := extensions.ValidateConstrainedAnyTypeConsistency(params, args, variadic)
+		require.NoError(t, err)
+	})
+
+	t.Run("variadic with mismatched types fails", func(t *testing.T) {
+		// func(any1, ...) with (i32, i64, i32)
```

Comment:
Maybe an extra test with `func(any1, any2, ...)`?

Similar to `func(any1, any2, any1, ...) with (i32, string, i64, i32)` below but not exactly the same?